In [78]:
# Imports required libraries and dependencies

import pandas as pd
import hvplot.pandas
from sklearn.cluster import KMeans
from sklearn.decomposition import PCA
from sklearn.preprocessing import StandardScaler
import matplotlib.pyplot as plt

In [79]:
# Loads the data into a Pandas DataFrame

df_market_data = pd.read_csv(
    "../Resources/crypto_market_data.csv",
    index_col="coin_id")

# Displays sample data

df_market_data.head(10)

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                    
bitcoin                              1.08388                     7.60278   
ethereum                             0.22392                    10.38134   
tether                              -0.21173                     0.04935   
ripple                              -0.37819                    -0.60926   
bitcoin-cash                         2.90585                    17.09717   
binancecoin                          2.10423                    12.85511   
chainlink                           -0.23935                    20.69459   
cardano                              0.00322                    13.99302   
litecoin                            -0.06341                     6.60221   
bitcoin-cash-sv                      0.92530                     3.29641   

                 price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                     
bitcoin                              6.57509                      7.67258   
ethereum                             4.80849                      0.13169   
tether                               0.00640                     -0.04237   
ripple                               2.24984                      0.23455   
bitcoin-cash                        14.75334                     15.74903   
binancecoin                          6.80688                      0.05865   
chainlink                            9.30098                    -11.21747   
cardano                              5.55476                     10.10553   
litecoin                             7.28931                      1.21662   
bitcoin-cash-sv                     -1.86656                      2.88926   

                 price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                      
bitcoin                             -3.25185                      83.51840   
ethereum                           -12.88890                     186.77418   
tether                               0.28037                      -0.00542   
ripple                             -17.55245                      39.53888   
bitcoin-cash                       -13.71793                      21.66042   
binancecoin                         36.33486                     155.61937   
chainlink                          -43.69522                     403.22917   
cardano                            -22.84776                     264.51418   
litecoin                           -17.23960                      27.49919   
bitcoin-cash-sv                    -24.87434                       7.42562   

                 price_change_percentage_1y  
coin_id                                      
bitcoin                            37.51761  
ethereum                          101.96023  
tether                              0.01954  
ripple                            -16.60193  
bitcoin-cash                       14.49384  
binancecoin                        69.69195  
chainlink                         325.13186  
cardano                           156.09756  
litecoin                          -12.66408  
bitcoin-cash-sv                    93.73082

In [80]:
# Generates summary statistics

df_market_data.describe()

price_change_percentage_24h  price_change_percentage_7d  \
count                    41.000000                   41.000000   
mean                     -0.269686                    4.497147   
std                       2.694793                    6.375218   
min                     -13.527860                   -6.094560   
25%                      -0.608970                    0.047260   
50%                      -0.063410                    3.296410   
75%                       0.612090                    7.602780   
max                       4.840330                   20.694590   

       price_change_percentage_14d  price_change_percentage_30d  \
count                    41.000000                    41.000000   
mean                      0.185787                     1.545693   
std                       8.376939                    26.344218   
min                     -18.158900                   -34.705480   
25%                      -5.026620                   -10.438470   
50%                       0.109740                    -0.042370   
75%                       5.510740                     4.578130   
max                      24.239190                   140.795700   

       price_change_percentage_60d  price_change_percentage_200d  \
count                    41.000000                     41.000000   
mean                     -0.094119                    236.537432   
std                      47.365803                    435.225304   
min                     -44.822480                     -0.392100   
25%                     -25.907990                     21.660420   
50%                      -7.544550                     83.905200   
75%                       0.657260                    216.177610   
max                     223.064370                   2227.927820   

       price_change_percentage_1y  
count                   41.000000  
mean                   347.667956  
std                   1247.842884  
min                    -17.567530  
25%                      0.406170  
50%                     69.691950  
75%                    168.372510  
max                   7852.089700

In [81]:
# Plots data to see what's in the DataFrame

df_market_data.hvplot.line(
    width=800,
    height=400,
    rot=90
)

:NdOverlay   [Variable]
   :Curve   [coin_id]   (value)

### Prepare the Data

In [7]:
# Use the `StandardScaler()` module from scikit-learn to normalize the data from the CSV file

scaler = StandardScaler()

scaler.fit(df_market_data)

normalized_data = scaler.transform(df_market_data)


In [82]:
# Creates a DataFrame with the scaled data

scaled_df = pd.DataFrame(normalized_data)


# Copies the crypto names from the original data

cryto_names = df_market_data.index

# Sets the coinid column as index for new DataFrame

scaled_df.index = cryto_names 


# Sets the columns of new DataFrame to the columns of the original DataFrame 

scaled_df.columns = df_market_data.columns


# Displays the scaled DataFrame

scaled_df

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                         
bitcoin                                  0.508529                    0.493193   
ethereum                                 0.185446                    0.934445   
tether                                   0.021774                   -0.706337   
ripple                                  -0.040764                   -0.810928   
bitcoin-cash                             1.193036                    2.000959   
binancecoin                              0.891871                    1.327295   
chainlink                                0.011397                    2.572251   
cardano                                  0.102530                    1.508001   
litecoin                                 0.077497                    0.334297   
bitcoin-cash-sv                          0.448952                   -0.190684   
crypto-com-chain                         0.331280                   -1.614844   
usd-coin                                 0.034352                   -0.733026   
eos                                      0.155710                   -0.922491   
monero                                   0.262723                    1.792602   
tron                                     0.130050                   -0.041018   
tezos                                   -0.151583                    0.708196   
okb                                     -0.923203                   -1.437359   
stellar                                 -0.277543                   -0.385209   
cosmos                                  -0.255978                    1.840274   
cdai                                     0.180851                   -0.704931   
neo                                      0.286546                   -0.326301   
wrapped-bitcoin                          0.515453                    0.461843   
leo-token                                0.051758                   -0.928381   
huobi-token                             -0.052032                   -0.457229   
nem                                     -0.217984                   -0.849381   
binance-usd                              0.061339                   -0.706669   
iota                                     0.259097                    0.249508   
vechain                                  0.585089                   -0.994231   
zcash                                   -0.127467                    0.929119   
theta-token                             -1.612188                   -1.682027   
dash                                    -0.296940                    0.094763   
ethereum-classic                        -0.071312                   -0.229484   
ethlend                                 -4.981042                   -0.045178   
maker                                   -0.125168                    0.580730   
havven                                  -1.428574                   -0.025510   
omisego                                  1.919812                    0.370447   
celsius-degree-token                     1.045530                   -0.618328   
ontology                                -0.409044                   -0.906963   
ftx-token                                0.414711                    0.414044   
true-usd                                 0.078038                   -0.687745   
digibyte                                 1.217453                   -0.607714   

                      price_change_percentage_14d  \
coin_id                                             
bitcoin                                  0.772200   
ethereum                                 0.558692   
tether                                  -0.021680   
ripple                                   0.249458   
bitcoin-cash                             1.760610   
binancecoin                              0.800214   
chainlink                                1.101647   
cardano                                  0.648885   
litecoin

---

### Find the Best Value for k Using the Original Data.

In [97]:
# Creates a list with the number of k-values from 1 to 11

k_values = list(range(1,12))

k_values

[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11]

In [98]:
# Creates an empty list to store the inertia values

inertia = []

# Creates a for loop to compute the inertia with each possible value of k
for each_k in k_values:
# Inside the loop:
    # 1. Creates a KMeans model using the loop counter for the n_clusters

    model = KMeans(n_clusters= each_k, random_state= 1)
    # 2. Fits the model to the data using `scaled_df`

    model.fit(scaled_df)
    # 3. Appends the model.inertia_ to the inertia list

    inertia.append(model.inertia_)

In [99]:
# Creates a dictionary with the data to plot the Elbow curve

elbow_data = {'k': k_values, 'inertia': inertia}


# Creates a DataFrame with the data to plot the Elbow curve

elbow_df = pd.DataFrame(elbow_data)

elbow_df

k     inertia
0    1  287.000000
1    2  212.123342
2    3  165.136752
3    4   79.022435
4    5   66.413051
5    6   61.634088
6    7   46.074467
7    8   43.392694
8    9   32.913223
9   10   30.469111
10  11   26.480101

In [110]:
# Plots a line chart with all the inertia values computed with the different values of k to visually identify the optimal value for k.

no_optimization_plot = elbow_df.hvplot.line(
                            x = "k",
                            y = "inertia",
                            title = "Elbow Curve",
                            xticks = "k"
                        )

no_optimization_plot

:Curve   [k]   (inertia)

#### Answer the following question: 

**Question:** What is the best value for `k`?

**Answer:** <font color = 'aquamarine'> The best value of k seems to be 4. Any k-value after 4 results in a stagnated inertia range, so 4 clusters looks like it should be the highest number of clusters that should be accepted for this data. </font>

---

### Cluster Cryptocurrencies with K-means Using the Original Data

In [101]:
# Initializes the K-Means model using the best value for k

model = KMeans(n_clusters= 4, random_state= 1)

In [102]:
# Fits the K-Means model using the scaled data

model.fit(scaled_df)

KMeans(n_clusters=4, random_state=1)

In [89]:
# Predicts the clusters to group the cryptocurrencies using the scaled data

cluster_prediction = model.predict(scaled_df)

# Prints the resulting array of cluster values.

print(cluster_prediction)

[2 2 0 0 2 2 2 2 2 0 0 0 0 2 0 2 0 0 2 0 0 2 0 0 0 0 0 0 2 0 0 0 3 2 0 0 1
 0 0 0 0]


In [90]:
# Creates a copy of the DataFrame

copied_scaled_df = scaled_df.copy()

copied_scaled_df

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                         
bitcoin                                  0.508529                    0.493193   
ethereum                                 0.185446                    0.934445   
tether                                   0.021774                   -0.706337   
ripple                                  -0.040764                   -0.810928   
bitcoin-cash                             1.193036                    2.000959   
binancecoin                              0.891871                    1.327295   
chainlink                                0.011397                    2.572251   
cardano                                  0.102530                    1.508001   
litecoin                                 0.077497                    0.334297   
bitcoin-cash-sv                          0.448952                   -0.190684   
crypto-com-chain                         0.331280                   -1.614844   
usd-coin                                 0.034352                   -0.733026   
eos                                      0.155710                   -0.922491   
monero                                   0.262723                    1.792602   
tron                                     0.130050                   -0.041018   
tezos                                   -0.151583                    0.708196   
okb                                     -0.923203                   -1.437359   
stellar                                 -0.277543                   -0.385209   
cosmos                                  -0.255978                    1.840274   
cdai                                     0.180851                   -0.704931   
neo                                      0.286546                   -0.326301   
wrapped-bitcoin                          0.515453                    0.461843   
leo-token                                0.051758                   -0.928381   
huobi-token                             -0.052032                   -0.457229   
nem                                     -0.217984                   -0.849381   
binance-usd                              0.061339                   -0.706669   
iota                                     0.259097                    0.249508   
vechain                                  0.585089                   -0.994231   
zcash                                   -0.127467                    0.929119   
theta-token                             -1.612188                   -1.682027   
dash                                    -0.296940                    0.094763   
ethereum-classic                        -0.071312                   -0.229484   
ethlend                                 -4.981042                   -0.045178   
maker                                   -0.125168                    0.580730   
havven                                  -1.428574                   -0.025510   
omisego                                  1.919812                    0.370447   
celsius-degree-token                     1.045530                   -0.618328   
ontology                                -0.409044                   -0.906963   
ftx-token                                0.414711                    0.414044   
true-usd                                 0.078038                   -0.687745   
digibyte                                 1.217453                   -0.607714   

                      price_change_percentage_14d  \
coin_id                                             
bitcoin                                  0.772200   
ethereum                                 0.558692   
tether                                  -0.021680   
ripple                                   0.249458   
bitcoin-cash                             1.760610   
binancecoin                              0.800214   
chainlink                                1.101647   
cardano                                  0.648885   
litecoin

In [91]:
# Adds a new column to the DataFrame with the predicted clusters

copied_scaled_df['predicted'] = cluster_prediction

# Displays sample data

copied_scaled_df

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                         
bitcoin                                  0.508529                    0.493193   
ethereum                                 0.185446                    0.934445   
tether                                   0.021774                   -0.706337   
ripple                                  -0.040764                   -0.810928   
bitcoin-cash                             1.193036                    2.000959   
binancecoin                              0.891871                    1.327295   
chainlink                                0.011397                    2.572251   
cardano                                  0.102530                    1.508001   
litecoin                                 0.077497                    0.334297   
bitcoin-cash-sv                          0.448952                   -0.190684   
crypto-com-chain                         0.331280                   -1.614844   
usd-coin                                 0.034352                   -0.733026   
eos                                      0.155710                   -0.922491   
monero                                   0.262723                    1.792602   
tron                                     0.130050                   -0.041018   
tezos                                   -0.151583                    0.708196   
okb                                     -0.923203                   -1.437359   
stellar                                 -0.277543                   -0.385209   
cosmos                                  -0.255978                    1.840274   
cdai                                     0.180851                   -0.704931   
neo                                      0.286546                   -0.326301   
wrapped-bitcoin                          0.515453                    0.461843   
leo-token                                0.051758                   -0.928381   
huobi-token                             -0.052032                   -0.457229   
nem                                     -0.217984                   -0.849381   
binance-usd                              0.061339                   -0.706669   
iota                                     0.259097                    0.249508   
vechain                                  0.585089                   -0.994231   
zcash                                   -0.127467                    0.929119   
theta-token                             -1.612188                   -1.682027   
dash                                    -0.296940                    0.094763   
ethereum-classic                        -0.071312                   -0.229484   
ethlend                                 -4.981042                   -0.045178   
maker                                   -0.125168                    0.580730   
havven                                  -1.428574                   -0.025510   
omisego                                  1.919812                    0.370447   
celsius-degree-token                     1.045530                   -0.618328   
ontology                                -0.409044                   -0.906963   
ftx-token                                0.414711                    0.414044   
true-usd                                 0.078038                   -0.687745   
digibyte                                 1.217453                   -0.607714   

                      price_change_percentage_14d  \
coin_id                                             
bitcoin                                  0.772200   
ethereum                                 0.558692   
tether                                  -0.021680   
ripple                                   0.249458   
bitcoin-cash                             1.760610   
binancecoin                              0.800214   
chainlink                                1.101647   
cardano                                  0.648885   
litecoin

In [111]:

# Creates a scatter plot using hvPlot by setting x = "price_change_percentage_24h" and y = "price_change_percentage_7d" 

no_optimization_scatter = copied_scaled_df.hvplot.scatter(
                                x = "price_change_percentage_24h",
                                y = "price_change_percentage_7d",
# Colors the graph points with the labels found using K-Means
                                by = "predicted",
# Adds the crypto name in the `hover_cols` parameter to identify the cryptocurrency represented by each data point
                                hover_cols = "coin_id"
                            )

no_optimization_scatter

:NdOverlay   [predicted]
   :Scatter   [price_change_percentage_24h]   (price_change_percentage_7d,coin_id)

---

### Optimize Clusters with Principal Component Analysis.

In [93]:
# Creates a PCA model instance and sets `n_components = 3`

pca = PCA(n_components= 3)


In [94]:
# Uses the PCA model with `fit_transform` to reduce to three principal components

component_data = pca.fit_transform(scaled_df)

# Views the first five rows  

component_data[:5]



array([[-0.60066733,  0.84276006,  0.46159457],
       [-0.45826071,  0.45846566,  0.95287678],
       [-0.43306981, -0.16812638, -0.64175193],
       [-0.47183495, -0.22266008, -0.47905316],
       [-1.15779997,  2.04120919,  1.85971527]])

In [95]:
# Retrieves the explained variance to determine how much information can be attributed to each principal component

variance_ratio = pca.explained_variance_ratio_

variance_sum = variance_ratio.sum()

print()
print(f"The explained variance is {variance_ratio}")
print()
print(f"The total explained variance is {variance_sum}")


The explained variance is [0.3719856  0.34700813 0.17603793]

The total explained variance is 0.895031657030984


#### Answer the following question: 

**Question:** What is the total explained variance of the three principal components?

**Answer:** <font color = "aquamarine"> The total explained variance of the three principal components is 0.895 </font>

In [96]:
# Creates a new DataFrame with the PCA data

pca_df = pd.DataFrame(component_data, columns= ["PCA1", "PCA2", "PCA3"])

# Copies the crypto names from the original data

original_cryto_names = df_market_data.index

# Sets the coinid column as index of new DataFrame

pca_df.index = original_cryto_names

# Displays the new DataFrame

pca_df.head()

PCA1      PCA2      PCA3
coin_id                                   
bitcoin      -0.600667  0.842760  0.461595
ethereum     -0.458261  0.458466  0.952877
tether       -0.433070 -0.168126 -0.641752
ripple       -0.471835 -0.222660 -0.479053
bitcoin-cash -1.157800  2.041209  1.859715

---

### Find the Best Value for k Using the PCA Data

In [64]:
# Create a list with the number of k-values from 1 to 11

k_list = list(range(1,11))

print(k_list)


[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11]


In [65]:
# Creates an empty list to store the inertia values

inertia_list = []

# Creates a for loop to compute the inertia with each possible value of k
for k in k_list:
    # Inside the loop:
    # 1. Creates a KMeans model using the loop counter for the n_clusters
    kmeans_model = KMeans(n_clusters= k)
    # 2. Fits the model to the data using `pca_df`
    kmeans_model.fit(pca_df)
    # 3. Appends the model.inertia_ to the inertia list
    inertia_list.append(kmeans_model.inertia_)


In [66]:
# Creates a dictionary with the data to plot the Elbow curve

pca_dictionary = {'k': k_list, 'inertia': inertia_list}

# Creates a DataFrame with the data to plot the Elbow curve

pca_elbow_df = pd.DataFrame(pca_dictionary)

pca_elbow_df

k     inertia
0    1  256.874086
1    2  168.811896
2    3  140.431833
3    4   49.665497
4    5   42.601476
5    6   32.923491
6    7   21.145312
7    8   18.832838
8    9   13.908202
9   10   13.200464
10  11    8.058822

In [113]:
# Plots a line chart with all the inertia values computed with the different values of k to visually identify the optimal value for k

optimization_plot = pca_elbow_df.hvplot.line(
                            x = 'k',
                            y = 'inertia',
                            title =  'Elbow Curve Using PCA',
                            xticks = 'k'
                        )

optimization_plot

:Curve   [k]   (inertia)

#### Answer the following questions: 

* **Question:** What is the best value for `k` when using the PCA data?

  * **Answer:** <font color = 'aquamarine'> The best value for k when using the PCA data is 4 </font>

<br>

* **Question:** Does it differ from the best k value found using the original data?

  * **Answer:**  <font color = 'aquamarine'> No, the k-value for this method was 4 just like it was for the original data. So, it does not seem to differ. </font>

### Cluster Cryptocurrencies with K-means Using the PCA Data

In [103]:
# Initializes the K-Means model using the best value for k

optimal_k_model  = KMeans(n_clusters= 4)


In [104]:
# Fits the K-Means model using the PCA data

optimal_k_model.fit(pca_df)

KMeans(n_clusters=4)

In [105]:
# Predicts the clusters to group the cryptocurrencies using the PCA data

clusters = optimal_k_model.predict(pca_df)

# Prints the resulting array of cluster values

print(clusters)

[0 0 2 2 0 0 0 0 0 2 2 2 2 0 2 0 2 2 0 2 2 0 2 2 2 2 2 2 0 2 2 2 1 0 2 2 3
 2 2 2 2]


In [106]:
# Creates a copy of the DataFrame with the PCA data

copied_pca_df = pca_df.copy()

# Adds a new column to the DataFrame with the predicted clusters

copied_pca_df['predictions'] = clusters

# Displays the DataFrame

copied_pca_df


PCA1      PCA2      PCA3  predictions
coin_id                                                        
bitcoin              -0.600667  0.842760  0.461595            0
ethereum             -0.458261  0.458466  0.952877            0
tether               -0.433070 -0.168126 -0.641752            2
ripple               -0.471835 -0.222660 -0.479053            2
bitcoin-cash         -1.157800  2.041209  1.859715            0
binancecoin          -0.516534  1.388377  0.804071            0
chainlink            -0.450711  0.517699  2.846143            0
cardano              -0.345600  0.729439  1.478013            0
litecoin             -0.649468  0.432165  0.600303            0
bitcoin-cash-sv      -0.759014 -0.201200 -0.217653            2
crypto-com-chain     -0.248198 -1.376252 -1.462026            2
usd-coin             -0.438408 -0.175337 -0.663388            2
eos                  -0.693425 -0.473815 -0.527597            2
monero                0.060499  2.909404  1.498571            0
tron                 -0.393352 -0.108192 -0.012756            2
tezos                -0.796176 -0.494409  1.082812            0
okb                   0.064075 -1.269825 -1.098829            2
stellar              -0.489015 -0.732719 -0.062543            2
cosmos               -0.306272  0.703415  1.714224            0
cdai                 -0.513528 -0.142802 -0.656566            2
neo                  -0.362120 -0.986914 -0.728752            2
wrapped-bitcoin      -0.604265  0.827398  0.439316            0
leo-token            -0.413296 -0.674115 -1.076628            2
huobi-token          -0.407483 -0.212507 -0.351426            2
nem                   0.608974  0.563532 -1.148742            2
binance-usd          -0.450211 -0.151019 -0.647401            2
iota                 -0.764665 -0.517886  0.204990            2
vechain              -0.556315 -1.938209 -1.261776            2
zcash                -0.425147  0.492976  1.058048            0
theta-token           2.676868 -0.013954 -1.965207            2
dash                 -0.613923 -0.479337  0.339565            2
ethereum-classic     -0.579924 -0.356334 -0.114942            2
ethlend               8.089018 -3.896891  2.301382            1
maker                -0.389045  0.165041  0.379414            0
havven                0.865762 -2.261882  0.275583            2
omisego               0.111675  0.428316 -1.205398            2
celsius-degree-token  4.792395  6.767679 -1.986985            3
ontology             -0.632355 -2.108117 -0.652227            2
ftx-token            -0.593142  0.021485  0.209911            2
true-usd             -0.458131 -0.135734 -0.635284            2
digibyte             -0.297910 -0.191126 -0.909602            2

In [114]:
# Creates a scatter plot using hvPlot by setting x="PCA1" and y="PCA2"

optimization_scatter = copied_pca_df.hvplot.scatter(
                                    x="PCA1",
                                    y="PCA2",
# Colors the graph points with the labels found using K-Means
                                    by = "predictions",
# Adds the crypto name in the `hover_cols` parameter to identify the cryptocurrency represented by each data point
                                    hover_cols = "coin_id"
                                )

optimization_scatter 


:NdOverlay   [predictions]
   :Scatter   [PCA1]   (PCA2,coin_id)

### Visualize and Compare the Results

In this section, you will visually analyze the cluster analysis results by contrasting the outcome with and without using the optimization techniques.

In [116]:
# Composite plot to contrast the Elbow curves

plots = (no_optimization_plot + optimization_plot).cols(1)

plots


:Layout
   .Curve.I  :Curve   [k]   (inertia)
   .Curve.II :Curve   [k]   (inertia)

In [117]:
# Composite plot to contrast the clusters

cluster_plots = (no_optimization_scatter + optimization_scatter).cols(1)

cluster_plots

:Layout
   .NdOverlay.I  :NdOverlay   [predicted]
      :Scatter   [price_change_percentage_24h]   (price_change_percentage_7d,coin_id)
   .NdOverlay.II :NdOverlay   [predictions]
      :Scatter   [PCA1]   (PCA2,coin_id)

#### Answer the following question: 

  * **Question:** After visually analyzing the cluster analysis results, what is the impact of using fewer features to cluster the data using K-Means?

  * **Answer:** 
  
  <font color = 'aquamarine'> According to the first composite plot, reducing the number of features does not appear to have an effect on the amount of clusters. Both the plots indicate that 4 is the best number of clusters to use for the model. The overall inertia of the PCA line graph looks like it is smaller compared to the non-optimized graph. This trend indicates that the clustering that occurred through PCA optimization had better clustering than the clustering that was done without optimization.    
  
  By looking at the second pair of graphs (the scatterplots), the clustering after reducing the dimensionality does not show signs of negatively impacting the clustering. It can even be argued that the clustering is better after the PCA procedure due to the smaller inertia values. </font>